rlkit/core/rl_algorithm.py

```diff
@@ -163,68 +150,74 @@ def train(self):
-                    self.collect_data(self.exploration_policy, explore=True, num_samples=self.max_path_length*10)
-
-                    # gathers more data conditioned on embedding from initial exploration trajectories
-                    self.set_policy_eval_z(self.task_idx, eval_task=False)
-                    self.collect_data(self.policy, explore=False, num_samples=self.max_path_length*10)
-                else:
-                    raise Exception("Invalid option for computing embedding")
-
-            # Sample data from train tasks.
-            for i in range(self.num_env_steps_per_epoch): # num iterations
-                idx = np.random.randint(len(self.train_tasks))
-                self.task_idx = idx
-                self.env.reset_task(idx)
-                # TODO: add flag for this
-                # self.collect_data(self.exploration_policy, explore=True, num_samples=self.max_path_length*10)
-                self.set_policy_eval_z(self.task_idx, eval_task=False)
-                self.collect_data(self.policy, explore=True, num_samples=self.max_path_length*10)
+
+                    if self.embedding_source == 'initial_pool':
+                        pass
+                    elif self.embedding_source == 'online_exploration_trajectories':
+                        self.eval_enc_replay_buffer.task_buffers[idx].clear()
+                        # resamples using current policy, conditioned on prior
+                        self.collect_data_sampling_from_prior(num_samples=self.num_steps_per_task, eval_task=True)
+                    elif self.embedding_source == 'online_on_policy_trajectories':
+                        # Clear the encoding replay buffer, so at eval time
+                        # We are computing z only from trajectories from the current epoch.
+                        self.eval_enc_replay_buffer.task_buffers[idx].clear()
+
+                        # regathers with online exploration trajectories
+                        self.collect_data_sampling_from_prior(num_samples=self.num_steps_per_task, eval_task=True)
+                        self.collect_data_from_task_posterior(idx=idx, num_samples=self.num_steps_per_task, eval_task=True)
+                    else:
+                        raise Exception("Invalid option for computing embedding")
+
+                # Sample data from train tasks.
+                for i in range(self.num_tasks_sample):
+                    idx = np.random.randint(len(self.train_tasks))
+                    self.task_idx = idx
+                    self.env.reset_task(idx)
+                    # TODO: add flag for this
+                    # TODO(KR) see todo at start of epoch about this
+                    # self.collect_data(self.exploration_policy, explore=True, num_samples=self.max_path_length*10)
+                    self.collect_data_from_task_posterior(idx=idx, num_samples=self.num_steps_per_task, eval_task=False)
```

Comment:
One thing that bothers me is that the way this currently is done, we sample one z, and use that single z to gather num_steps_per_task. We probably want to be able to periodically resample z, both when sampling from the prior and posterior, which may just make this rather ugly.